django/db/models/sql/compiler.py

```diff
@@ -1289,25 +1291,26 @@ def as_sql(self):
                 for p, vals in zip(placeholder_rows, param_rows)
             ]
 
-    def execute_sql(self, return_id=False):
-        assert not (
-            return_id and len(self.query.objs) != 1 and
-            not self.connection.features.can_return_ids_from_bulk_insert
+    def execute_sql(self):
+        return_id = (
+            len(self.query.objs) == 1 or
+            self.connection.features.can_return_rows_from_bulk_insert
         )
-        self.return_id = return_id
         with self.connection.cursor() as cursor:
             for sql, params in self.as_sql():
                 cursor.execute(sql, params)
-            if not return_id:
-                return
-            if self.connection.features.can_return_ids_from_bulk_insert and len(self.query.objs) > 1:
-                return self.connection.ops.fetch_returned_insert_ids(cursor)
-            if self.connection.features.can_return_id_from_insert:
+            if not (return_id and self.query.get_meta().returning):
+                return []
+            if self.connection.features.can_return_rows_from_bulk_insert and len(self.query.objs) > 1:
+                return self.connection.ops.fetch_returned_insert_rows(cursor)
+            if self.connection.features.can_return_columns_from_insert:
                 assert len(self.query.objs) == 1
-                return self.connection.ops.fetch_returned_insert_id(cursor)
-            return self.connection.ops.last_insert_id(
-                cursor, self.query.get_meta().db_table, self.query.get_meta().pk.column
-            )
+                return self.connection.ops.fetch_returned_insert_columns(cursor)
+            if isinstance(self.query.get_meta().pk, AutoField):
```

Comment:
Do we need the check for `AutoField` or can we just return `last_insert_id` unconditionally like before?

Reply:
Yes, we do for MySQL and SQLite in cases where people chose custom primary keys that are not AI (auto increment) integer fields.

Reply:
Maybe I add a comment ;)

Reply:
Maybe we need a field property. It doesn't look nice to put `isinstance()` special cases here.